firestore/src/jni/arena_ref.h

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIREBASE_FIRESTORE_SRC_JNI_ARENA_REF_H_
+#define FIREBASE_FIRESTORE_SRC_JNI_ARENA_REF_H_
+
+#include <jni.h>
+
+#include <cstdlib>
+
+#include "firestore/src/jni/jni_fwd.h"
+#include "firestore/src/jni/object.h"
+#include "firestore/src/jni/ownership.h"
+
+namespace firebase {
+namespace firestore {
+namespace jni {
+
+/**
+ * ArenaRef is an RAII wrapper which serves the same purpose as Global<T>, both
+ * of them manage the lifetime of JNI reference. Compared to Global<T>, the
+ * advantage of ArenaRef is it gets rid of the restriction of total number of
+ * JNI global reference (the pointer inside Global<T>) need to be under 51,200,
+ * since after that number the app will crash. The drawback is ArenaRef holds
+ * JNI Reference indirectly using HashMap so there is a efficiency lost against
+ * Global<T>.
+ */
+class ArenaRef {
+ public:
+  ArenaRef() = default;
+  ArenaRef(Env&, const Object&);
+  ArenaRef(const ArenaRef& other);
+  ArenaRef(ArenaRef&& other);
+  ArenaRef& operator=(const ArenaRef& other);
+  ArenaRef& operator=(ArenaRef&& other);
+  ~ArenaRef();
+
+  static void Initialize(Env&);
+
+  Local<Object> get(Env&) const;
+
+ private:
+  Local<Long> key_object(Env&) const;
+
+  int64_t key_ = -1;
```

Comment:
The more I think about it, the more I don't like this "magic value" of -1 (formerly 0) for `key_`. IMO it detracts from readability and maintainability where you see this comparison to -1 sprinkled through the code. I think a better solution is to assign `key_` once in the constructor, and then never change it. The exception to this rule would be in the move constructor and assignment operator, which could just use `std::swap` on the key_.). Then, use a boolean flag `valid_` which is true only if `key_` maps to an entry in the HashMap. What do you think?

Reply:
I got curious and tried it out this morning and it seems to work perfectly.
https://gist.github.com/dconeybe/afdcd15cd56bebe5497c07cbb8ceb935

There was just one unit test that failed, but it was assuming that the moved-from object was "null". Unless we make some strong guarantees about the state of the moved-from object, we should probably not test anything about the moved-from object, except its destructor and re-assigning it using the operator= overloads. Formally, those are the only valid operations guaranteed by the C++ standard to perform on a moved-from object.

Reply:
(I just realized that in that gist link I forgot to set `valid_ = other.valid_` in the copy assignment operator).

Reply:
I agree we shouldn't test the object after moving.

However, I think introduced another flag `valid` decrease the readability and maintainability. What I think is better is add MACRO like `invalid key` and set it to -1.

Reply:
Ok fair enough. There is definitely an argument for readability either way.